src/main/java/bio/terra/common/kubernetes/KubeService.java

```diff
@@ -0,0 +1,195 @@
+   */
+  public Set<String> getPodList() {
+    Set<String> pods = new HashSet<>();
+    if (!inKubernetes) {
+      return pods;
+    }
+
+    try {
+      CoreV1Api api = makeCoreApi();
+      V1PodList list =
+          api.listNamespacedPod(namespace, null, null, null, null, null, null, null, null, null);
+      for (V1Pod item : list.getItems()) {
+        if (item.getMetadata() != null) {
+          String podName = item.getMetadata().getName();
+          if (StringUtils.contains(podName, podNameFilter)) {
+            pods.add(podName);
+          }
+        }
+      }
+      return pods;
+    } catch (ApiException ex) {
+      throw new KubeApiException("Error listing pods", ex);
+    }
+  }
+
+  /**
+   * Launch the {@link KubePodListener} thread.
+   *
+   * @param stairway the Stairway instance passed to the pod listener for Stairway recovery
+   */
+  public void startPodListener(Stairway stairway) {
+    if (inKubernetes) {
+      podListener = new KubePodListener(this, stairway, namespace, podNameFilter);
+      podListenerThread = new Thread(podListener);
+      podListenerThread.start();
+    }
+  }
+
+  /**
+   * Stop the pod listener thread within a given time span.
+   *
+   * @param timeUnit unit of the joinWait
+   * @param joinWait number of units to wait for the listener thread to stop
+   * @return true if the thread joined in the time span. False otherwise.
+   */
+  public boolean stopPodListener(TimeUnit timeUnit, long joinWait) {
+    if (inKubernetes) {
+      podListenerThread.interrupt();
+      long waitMillis = timeUnit.toMillis(joinWait);
+      try {
+        podListenerThread.join(waitMillis);
+      } catch (InterruptedException ex) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public int getActivePodCount() {
+    if (podListener != null) {
+      return podListener.getActivePodCount();
+    }
+    int defaultPodCount = 1;
+    logger.info("KubeService ActivePodCount - default val: {}", defaultPodCount);
+    return defaultPodCount;
+  }
+
+  public void clearShutdown() {
+    isShutdown.set(false);
+  }
+
+  public void setShutdown() {
+    isShutdown.set(true);
+  }
+
+  public boolean isShutdown() {
+    return isShutdown.get();
+  }
+
+  public String getPodName() {
+    return podName;
+  }
+
+  public String getNamespace() {
+    return namespace;
+  }
+
+  private CoreV1Api makeCoreApi() {
+    try {
+      ApiClient client = ClientBuilder.cluster().build();
+      Configuration.setDefaultApiClient(client);
+      return new CoreV1Api();
+    } catch (IOException ex) {
+      throw new KubeApiException("Error making core API", ex);
+    }
+  }
+
+  private String readFileIntoString(String path) {
+    try {
+      byte[] encoded = Files.readAllBytes(Paths.get(path));
+      return new String(encoded, StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      logger.error("Failed to read file: " + path + "; ", e);
+      return null;
```

Comment:
Client may not be aware `getNamespace()` may return null. Feel it is possible to cause NPE in client code. Suggest to use `Optional`, or throw exception there?

Reply:
Did you push you commit to address this?

Reply:
Not yet. I should have this updated by days end.